plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergStatistics.java

```diff
@@ -0,0 +1,328 @@
+                                id,
+                                trinoType,
+                                trinoValue,
+                                trinoValue,
+                                Optional.of(0L),
+                                dataFile.recordCount());
+                        updateNullCountStats(id, Optional.of(0L));
+                    }
+                    else {
+                        // Update null counts, but do not clear min/max
+                        updateNullCountStats(id, Optional.of(dataFile.recordCount()));
+                    }
+                }
+                else {
+                    Object lowerBound = lowerBounds == null ? null : convertIcebergValueToTrino(column.type(), lowerBounds.get(id));
+                    Object upperBound = upperBounds == null ? null : convertIcebergValueToTrino(column.type(), upperBounds.get(id));
+                    Optional<Long> nullCount = Optional.ofNullable(dataFile.nullValueCounts().get(id));
+                    updateMinMaxStats(
+                            id,
+                            trinoType,
+                            lowerBound,
+                            upperBound,
+                            nullCount,
+                            dataFile.recordCount());
+                    updateNullCountStats(
+                            id,
+                            nullCount);
+                }
+            }
+        }
+
+        public IcebergStatistics build()
+        {
+            ImmutableMap.Builder<Integer, Object> minValues = ImmutableMap.builder();
+            ImmutableMap.Builder<Integer, Object> maxValues = ImmutableMap.builder();
+
+            columnStatistics.forEach((fieldId, statistics) -> {
+                statistics.getMin().ifPresent(min -> minValues.put(fieldId, min));
+                statistics.getMax().ifPresent(max -> maxValues.put(fieldId, max));
+            });
+
+            Map<Integer, Long> nullCounts = this.nullCounts.entrySet().stream()
+                    .filter(entry -> entry.getValue().isPresent())
+                    .collect(toImmutableMap(Map.Entry::getKey, entry -> entry.getValue().orElseThrow()));
+
+            return new IcebergStatistics(
+                    recordCount,
+                    fileCount,
+                    size,
+                    minValues.build(),
+                    maxValues.build(),
+                    nullCounts,
+                    ImmutableMap.copyOf(columnSizes));
+        }
+
+        private void updateNullCountStats(int id, Optional<Long> nullCount)
+        {
+            // If one file is missing nullCounts for a column, invalidate the estimate
+            if (nullCounts.containsKey(id) && nullCounts.get(id).isEmpty()) {
+                return;
+            }
+
+            if (nullCount.isPresent()) {
+                nullCounts.merge(id, nullCount, (existingCount, newCount) -> Optional.of(existingCount.get() + newCount.get()));
```

Comment:
You're doing 3 map lookups (contains, get, merge).
You can do this in one shot:

```
nullCounts.merge(id, nullCount, (existingCount, newCount) ->
        existingCount.isPresent() && newCount.isPresent() ? Optional.of(existingCount.get() + newCount.get()) : Optional.empty());
```

(you can extract lambda body to a method like `sumOfOptionals`)

Reply:
This method is also equivalent to this:
```java
nullCounts.merge(id, nullCount, (oldCount, newCount) -> 
        oldCount.flatMap(oldValue ->
                newCount.map(newValue -> newValue + oldValue)));

```
Though after looking at this, it's definitely less clear about the intent, though I would use this form if extracting a method to add optionals.

Actually, instead of just addition, maybe `mergeOptionals` would be better (I see somewhere else in this file it could be used). Here, it would be used like `mergeOptionals(oldCount, newCount, Long::sum)`.

```java
/**
 * Apply a function to the values in two optionals, returning an optional containing the result.
 * If either argument is empty, return empty.
 */
<A, B, C> Optional<C> mergeOptionals(Optional<A> a, Optional<B> b, BiFunction<A, B, C> mergeFunction)
{
    return a.flatMap(aa -> b.map(bb -> mergeFunction.apply(aa, bb)));
}
```

Reply:
>     return a.flatMap(aa -> b.map(bb -> mergeFunction.apply(aa, bb)));

i thought about that, but i don't find it _readable_, that's why i suggested `?:` use